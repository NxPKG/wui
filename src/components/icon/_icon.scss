.wuiIcon {
  flex-shrink: 0; // Ensures it never scales down below its intended size
  display: inline-block;
  vertical-align: middle;
  fill: currentColor;
  transform: translate(0, 0); // Hack to fix Firefox "softness"

  &:focus {
    opacity: 1; // We often hide icons on hover. Make sure they appear on focus.
    background: $wuiFocusBackgroundColor;
  }
}

// App icons are two-toned so get an extra class. This provides the base color.
.wuiIcon--app {
  fill: $wuiTextColor;

  // This provides the default secondary color
  .wuiIcon__fillSecondary {
    fill: makeGraphicContrastColor($wuiColorPrimary);
  }
}

.wuiLogoFill {
  fill: $wuiWazuhLogoPrimaryColor;
}

.wuiLogoSecondaryFill {
  fill: $wuiWazuhLogoSecondaryColor;
}

.wuiIcon-isLoading {
  opacity: $wuiIconLoadingOpacity;
  background-color: currentColor;
  border-radius: $wuiBorderRadius;
}

.wuiIcon-isLoaded {
  animation: wuiIconLoading $wuiAnimSpeedNormal ease-in 0s 1 forwards;
}

// COLORS

@each $name, $color in $wuiIconColors {
  .wuiIcon--#{$name} {
    // Use color attribute so currentColor can be used in more properties
    color: $color;
  }
}

// App icon special cases for fully mono
.wuiIcon--text,
.wuiIcon--subdued,
.wuiIcon--primary {
  &,
  .wuiIcon__fillSecondary {
    fill: currentColor;
  }
}

// Use the negative fill on SVG elements that need to flip portions
// of the icon to light and dark compatible modes. You can see this
// in use on the Wazuh logo icons.
.wuiIcon__fillNegative {
  fill: $wuiColorDarkestShade;
}

// SIZES

@each $name, $size in $wuiIconSizes {
  .wuiIcon--#{$name} {
    @include size($size);
  }
}

@keyframes wuiIconLoading {
  0% {
    opacity: $wuiIconLoadingOpacity;
  }

  100% {
    opacity: 1;
  }
}
