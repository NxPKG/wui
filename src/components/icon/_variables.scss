$wuiIconLoadingOpacity: .05;

// Modifier naming and colors.
$wuiIconColors: (
  accent: makeGraphicContrastColor($wuiColorAccent),
  danger: makeGraphicContrastColor($wuiColorDanger),
  ghost: $wuiColorGhost,
  primary: makeGraphicContrastColor($wuiColorPrimary),
  secondary: makeGraphicContrastColor($wuiColorSecondary),
  success: makeGraphicContrastColor($wuiColorSuccess),
  subdued: $wuiTextSubduedColor,
  text: $wuiTextColor,
  warning: makeGraphicContrastColor($wuiColorWarning),
);

$wuiIconSizes: (
  small: $wuiSizeM,
  medium: $wuiSize,
  large: $wuiSizeL,
  xLarge: $wuiSizeXL,
  xxLarge: $wuiSizeXXL
);
