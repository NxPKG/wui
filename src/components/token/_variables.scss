// sass-lint:disable indentation

$wuiTokenGrayColor: lightOrDarkTheme($wuiColorDarkShade, $wuiColorMediumShade);

// Appends the gray color to the color blind palette
$wuiTokenTypes: map-merge($wuiPaletteColorBlind, (
  'gray': (
    graphic: $wuiTokenGrayColor,
    behindText: $wuiTokenGrayColor,
  )
));

$wuiTokenTypeKeys: map-keys($wuiTokenTypes);
