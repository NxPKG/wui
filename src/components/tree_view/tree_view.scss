.wuiTreeView__wrapper .wuiTreeView {
  margin: 0;
  list-style-type: none;
}

.wuiTreeView .wuiTreeView {
  padding-left: $wuiSizeL;
}

.wuiTreeView__node {
  max-height: $wuiSizeXL;
  overflow: hidden;
  cursor: pointer;
  line-height: $wuiSizeXL;
}

.wuiTreeView__node--expanded {
  max-height: 100vh;
  overflow: auto;
}

.wuiTreeView__nodeInner {
  @include wuiTextTruncate;

  padding-left: $wuiSizeS;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: $wuiSizeXL;
  border-radius: $wuiBorderRadius;
  width: 100%;
  text-align-last: left;

  &:focus {
    box-shadow: inset 0 0 0 $wuiSizeXS / 4 $wuiFocusRingColor;
  }

  &:hover,
  &:active {
    background-color: tintOrShade($wuiColorLightShade, 50%, 10%);
  }

  .wuiTreeView__iconPlaceholder {
    width: $wuiSizeXL;
  }

}

.wuiTreeView__nodeLabel {
  @include wuiTextTruncate;
}

.wuiTreeView__iconWrapper {
  margin-top: -($wuiSizeXS / 2);
  margin-right: $wuiSizeS;

  // This helps tokens appear vertically centered
  .wuiToken {
    margin-top: $wuiSizeXS / 2;
  }
}

.wuiTreeView--compressed {
  .wuiTreeView__node {
    max-height: $wuiSizeL;
    line-height: $wuiSizeL;

    .wuiTreeView__nodeInner {
      height: $wuiSizeL;

    }

    .wuiTreeView__iconWrapper {
      margin: 0 ($wuiSizeS * .75) 0 0;
    }

    .wuiTreeView__nodeLabel {
      margin-top: -1px;
    }

    .wuiTreeView__iconPlaceholder {
      width: $wuiSizeL;
    }
  }

  .wuiTreeView__node--expanded {
    max-height: 100vh;
    overflow: auto;
  }
}

.wuiTreeView--withArrows {
  .wuiTreeView__expansionArrow {
    margin-right: $wuiSizeXS;
  }

  &.wuiTreeView {
    .wuiTreeView__nodeInner--withArrows {
      .wuiTreeView__iconWrapper {
        margin-left: 0;
      }
    }

    .wuiTreeView__iconWrapper {
      margin-left: $wuiSize + $wuiSizeXS;
    }
  }

  &.wuiTreeView--compressed {
    .wuiTreeView__nodeInner--withArrows {
      .wuiTreeView__iconWrapper {
        margin-left: 0;
      }
    }

    .wuiTreeView__iconWrapper {
      margin-left: $wuiSize;
    }
  }
}
