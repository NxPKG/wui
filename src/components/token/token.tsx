/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { FunctionComponent, HTMLAttributes } from 'react';
import { defaults } from 'lodash';
import classNames from 'classnames';
import { CommonProps, keysOf } from '../common';
import { isColorDark, hexToRgb } from '../../services';

import { IconType, WuiIcon, IconSize } from '../icon';
import { WuiTokenMapType, TOKEN_MAP } from './token_map';

type TokenSize = 'xs' | 's' | 'm' | 'l';
type TokenShape = 'circle' | 'square' | 'rectangle';
type TokenFill = 'dark' | 'light' | 'none';
type TokenColor =
  | 'wuiColorVis0'
  | 'wuiColorVis1'
  | 'wuiColorVis2'
  | 'wuiColorVis3'
  | 'wuiColorVis4'
  | 'wuiColorVis5'
  | 'wuiColorVis6'
  | 'wuiColorVis7'
  | 'wuiColorVis8'
  | 'wuiColorVis9'
  | 'gray';

const sizeToClassMap: { [size in TokenSize]: string } = {
  xs: 'wuiToken--xsmall',
  s: 'wuiToken--small',
  m: 'wuiToken--medium',
  l: 'wuiToken--large',
};

export const SIZES = keysOf(sizeToClassMap);

const shapeToClassMap: { [shape in TokenShape]: string } = {
  circle: 'wuiToken--circle',
  square: 'wuiToken--square',
  rectangle: 'wuiToken--rectangle',
};

export const SHAPES = keysOf(shapeToClassMap);

const fillToClassMap: { [fill in TokenFill]: string | null } = {
  none: null,
  light: 'wuiToken--light',
  dark: 'wuiToken--dark',
};

export const FILLS = keysOf(fillToClassMap);

const colorToClassMap: { [color in TokenColor]: string } = {
  wuiColorVis0: 'wuiToken--wuiColorVis0',
  wuiColorVis1: 'wuiToken--wuiColorVis1',
  wuiColorVis2: 'wuiToken--wuiColorVis2',
  wuiColorVis3: 'wuiToken--wuiColorVis3',
  wuiColorVis4: 'wuiToken--wuiColorVis4',
  wuiColorVis5: 'wuiToken--wuiColorVis5',
  wuiColorVis6: 'wuiToken--wuiColorVis6',
  wuiColorVis7: 'wuiToken--wuiColorVis7',
  wuiColorVis8: 'wuiToken--wuiColorVis8',
  wuiColorVis9: 'wuiToken--wuiColorVis9',
  gray: 'wuiToken--gray',
};

export const COLORS = keysOf(colorToClassMap);

export interface TokenProps {
  /**
   * An WUI icon type
   */
  iconType: IconType;
  /**
   * For best results use one of the vis color names (or 'gray').
   * Or supply your own color (can be used with dark or no fill only).
   * Default: `gray`
   */
  color?: TokenColor | string;
  /**
   * Outer shape surrounding the icon
   * Default: `circle`
   */
  shape?: TokenShape;
  /**
   * `light` for lightened color with border, `dark` for solid, or `none`
   * Default: `light`
   */
  fill?: TokenFill;
  /**
   * Size of the token
   */
  size?: TokenSize;
  /**
   * The icon's title. Required for accessibility
   */
  title?: string;
  'aria-label'?: string;
  'aria-labelledby'?: string;
  'aria-describedby'?: string;
}

export type WuiTokenProps = CommonProps &
  TokenProps &
  Omit<HTMLAttributes<HTMLSpanElement>, 'title'>;

export const WuiToken: FunctionComponent<WuiTokenProps> = ({
  iconType,
  color,
  fill,
  shape,
  size = 's',
  style = {},
  className,
  title,
  'aria-label': ariaLabel,
  'aria-labelledby': ariaLabelledby,
  'aria-describedby': ariaDescribedby,
  ...rest
}) => {
  // Set the icon size to the same as the passed size
  // unless they passed `xs` which IconSize doesn't support
  let finalSize: IconSize = size === 'xs' ? 's' : size;

  // When displaying at the small size, the token specific icons
  // should actually be displayed at medium size
  if (
    typeof iconType === 'string' &&
    iconType.indexOf('token') === 0 &&
    size === 's'
  ) {
    finalSize = 'm';
  }

  const currentDisplay = {
    color,
    fill,
    shape,
  };
  let finalDisplay;

  // If the iconType passed is one of the prefab token types,
  // grab its properties
  if (iconType in TOKEN_MAP) {
    const tokenDisplay = TOKEN_MAP[iconType as WuiTokenMapType];
    finalDisplay = defaults(currentDisplay, tokenDisplay);
  } else {
    finalDisplay = currentDisplay;
  }

  const finalColor = finalDisplay.color || 'gray';
  const finalShape = finalDisplay.shape || 'circle';
  let finalFill = finalDisplay.fill || 'light';

  // Color can be a named space via wuiColorVis
  let colorClass;
  if (finalColor in colorToClassMap) {
    colorClass = colorToClassMap[finalColor as TokenColor];
  }
  // Or it can be a string which adds inline styles for the
  else {
    // text color if fill='none' or
    if (finalFill === 'none') {
      style.color = finalColor;
    }
    // full background color if fill='dark' and overrides fill='light' with dark
    else {
      finalFill = 'dark';
      style.backgroundColor = finalColor;
      style.color = isColorDark(...hexToRgb(finalColor))
        ? '#FFFFFF'
        : '#000000';
    }
  }

  const classes = classNames(
    'wuiToken',
    colorClass,
    shapeToClassMap[finalShape],
    fillToClassMap[finalFill],
    sizeToClassMap[size],
    className
  );

  return (
    <span className={classes} style={style} {...rest}>
      <WuiIcon
        type={iconType}
        size={finalSize}
        title={title}
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledby}
        aria-describedby={ariaDescribedby}
      />
    </span>
  );
};
