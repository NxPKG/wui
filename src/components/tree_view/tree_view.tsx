/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { Component, HTMLAttributes, createContext } from 'react';
import classNames from 'classnames';
import { CommonProps } from '../common';
import { WuiI18n } from '../i18n';
import { WuiIcon } from '../icon';
import { WuiScreenReaderOnly } from '../accessibility';
import { WuiText } from '../text';
import { keys, htmlIdGenerator } from '../../services';

const WuiTreeViewContext = createContext<string>('');
const treeIdGenerator = htmlIdGenerator('wuiTreeView');

function hasAriaLabel(
  x: HTMLAttributes<HTMLUListElement>
): x is { 'aria-label': string } {
  return x.hasOwnProperty('aria-label');
}

export interface Node {
  /** An array of WuiTreeViewNodes to render as children
   */
  children?: Node[];
  /** The readable label for the item
   */
  label: string;
  /** A unique ID
   */
  id: string;
  /** An icon to use on the left of the label
   */
  icon?: React.ReactElement;
  /** Display a different icon when the item is expanded.
  For instance, an open folder or a down arrow
  */
  iconWhenExpanded?: React.ReactElement;
  /** Use an empty icon to keep items without an icon
  lined up with their siblings
  */
  useEmptyIcon?: boolean;
  /** Whether or not the item is expanded.
   */
  isExpanded?: boolean;
  /** Optional class to throw on the node
   */
  className?: string;
  /** Function to call when the item is clicked.
   The open state of the item will always be toggled.
   */
  callback?(): string;
}

export type WuiTreeViewDisplayOptions = 'default' | 'compressed';

const displayToClassNameMap: {
  [option in WuiTreeViewDisplayOptions]: string | null;
} = {
  default: null,
  compressed: 'wuiTreeView--compressed',
};

interface WuiTreeViewState {
  openItems: string[];
  activeItem: string;
  treeID: string;
  expandChildNodes: boolean;
}

export type CommonTreeProps = CommonProps &
  HTMLAttributes<HTMLUListElement> & {
    /** An array of WuiTreeViewNodes
     */
    items: Node[];
    /** Optionally use a variation with smaller text and icon sizes
     */
    display?: WuiTreeViewDisplayOptions;
    /** Set all items to open on initial load
     */
    expandByDefault?: boolean;
    /** Display expansion arrows next to all items
     * that contain children
     */
    showExpansionArrows?: boolean;
  };

export type WuiTreeViewProps = Omit<
  CommonTreeProps,
  'aria-label' | 'aria-labelledby'
> &
  ({ 'aria-label': string } | { 'aria-labelledby': string });

export class WuiTreeView extends Component<WuiTreeViewProps, WuiTreeViewState> {
  static contextType = WuiTreeViewContext;
  isNested: boolean = !!this.context;
  state: WuiTreeViewState = {
    openItems: this.props.expandByDefault
      ? this.props.items
          .map<string>(({ id, children }) =>
            children ? id : ((null as unknown) as string)
          )
          .filter(x => x != null)
      : this.props.items
          .map<string>(({ id, children, isExpanded }) =>
            children && isExpanded ? id : ((null as unknown) as string)
          )
          .filter(x => x != null),
    activeItem: '',
    treeID: this.context || treeIdGenerator(),
    expandChildNodes: this.props.expandByDefault || false,
  };

  buttonRef: Array<HTMLButtonElement | undefined> = [];

  setButtonRef = (
    ref: HTMLButtonElement | HTMLAnchorElement | null,
    index: number
  ) => {
    this.buttonRef[index] = ref as HTMLButtonElement;
  };

  handleNodeClick = (node: Node, ignoreCallback: boolean = false) => {
    const index = this.state.openItems.indexOf(node.id);

    this.setState({
      expandChildNodes: false,
    });

    node.isExpanded = !node.isExpanded;

    if (!ignoreCallback && node.callback !== undefined) {
      node.callback();
    }

    if (this.isNodeOpen(node)) {
      // if the node is part of openItems[] then remove it
      this.setState({
        openItems: this.state.openItems.filter((_, i) => i !== index),
      });
    } else {
      // if the node isn't part of openItems[] then add it
      this.setState(prevState => ({
        openItems: [...prevState.openItems, node.id],
        activeItem: node.id,
      }));
    }
  };

  // check if the node is included in openItems[]
  isNodeOpen = (node: Node) => {
    return this.state.openItems.includes(node.id);
  };

  // Enable keyboard navigation
  onKeyDown = (event: React.KeyboardEvent, node: Node) => {
    switch (event.key) {
      case keys.ARROW_DOWN: {
        const nodeButtons = Array.from(
          document.querySelectorAll(
            `[data-test-subj="wuiTreeViewButton-${this.state.treeID}"]`
          )
        );
        const currentIndex = nodeButtons.indexOf(event.currentTarget);
        if (currentIndex > -1) {
          const nextButton = nodeButtons[currentIndex + 1] as HTMLElement;
          if (nextButton) {
            event.preventDefault();
            event.stopPropagation();
            nextButton.focus();
          }
        }
        break;
      }
      case keys.ARROW_UP: {
        const nodeButtons = Array.from(
          document.querySelectorAll(
            `[data-test-subj="wuiTreeViewButton-${this.state.treeID}"]`
          )
        );
        const currentIndex = nodeButtons.indexOf(event.currentTarget);
        if (currentIndex > -1) {
          const prevButton = nodeButtons[currentIndex + -1] as HTMLElement;
          if (prevButton) {
            event.preventDefault();
            event.stopPropagation();
            prevButton.focus();
          }
        }
        break;
      }
      case keys.ARROW_RIGHT: {
        if (!this.isNodeOpen(node)) {
          event.preventDefault();
          event.stopPropagation();
          this.handleNodeClick(node, true);
        }
        break;
      }
      case keys.ARROW_LEFT: {
        if (this.isNodeOpen(node)) {
          event.preventDefault();
          event.stopPropagation();
          this.handleNodeClick(node, true);
        }
      }
      default:
        break;
    }
  };

  onChildrenKeydown = (event: React.KeyboardEvent, index: number) => {
    if (event.key === keys.ARROW_LEFT) {
      event.preventDefault();
      event.stopPropagation();
      this.buttonRef[index]!.focus();
    }
  };

  render() {
    const {
      children,
      className,
      items,
      display = 'default',
      expandByDefault,
      showExpansionArrows,
      ...rest
    } = this.props;

    // Computed classNames
    const classes = classNames(
      'wuiTreeView',
      display ? displayToClassNameMap[display] : null,
      { 'wuiTreeView--withArrows': showExpansionArrows },
      className
    );

    const instructionsId = `${this.state.treeID}--instruction`;

    return (
      <WuiTreeViewContext.Provider value={this.state.treeID}>
        <WuiText
          size={display === 'compressed' ? 's' : 'm'}
          className="wuiTreeView__wrapper">
          {!this.isNested && (
            <WuiI18n
              token="wuiTreeView.listNavigationInstructions"
              default="You can quickly navigate this list using arrow keys.">
              {(listNavigationInstructions: string) => (
                <WuiScreenReaderOnly>
                  <p id={instructionsId}>{listNavigationInstructions}</p>
                </WuiScreenReaderOnly>
              )}
            </WuiI18n>
          )}
          <ul
            className={classes}
            id={this.state.treeID}
            aria-describedby={!this.isNested ? instructionsId : undefined}
            {...rest}>
            {items.map((node, index) => {
              const buttonId = `${this.state.treeID}--${index}--node`;

              return (
                <WuiI18n
                  key={node.label + index}
                  token="wuiTreeView.ariaLabel"
                  default="{nodeLabel} child of {ariaLabel}"
                  values={{
                    nodeLabel: node.label,
                    ariaLabel: hasAriaLabel(rest) ? rest['aria-label'] : '',
                  }}>
                  {(ariaLabel: string) => {
                    const label:
                      | { 'aria-label': string }
                      | { 'aria-labelledby': string } = hasAriaLabel(rest)
                      ? {
                          'aria-label': ariaLabel,
                        }
                      : {
                          'aria-labelledby': `${buttonId} ${rest['aria-labelledby']}`,
                        };

                    const nodeClasses = classNames(
                      'wuiTreeView__node',
                      display ? displayToClassNameMap[display] : null,
                      { 'wuiTreeView__node--expanded': this.isNodeOpen(node) }
                    );

                    const nodeButtonClasses = classNames(
                      'wuiTreeView__nodeInner',
                      showExpansionArrows && node.children
                        ? 'wuiTreeView__nodeInner--withArrows'
                        : null,
                      this.state.activeItem === node.id
                        ? 'wuiTreeView__node--active'
                        : null,
                      node.className ? node.className : null
                    );

                    return (
                      <React.Fragment>
                        <li className={nodeClasses}>
                          <button
                            id={buttonId}
                            aria-controls={`wuiNestedTreeView-${this.state.treeID}`}
                            aria-expanded={this.isNodeOpen(node)}
                            ref={ref => this.setButtonRef(ref, index)}
                            data-test-subj={`wuiTreeViewButton-${this.state.treeID}`}
                            onKeyDown={(event: React.KeyboardEvent) =>
                              this.onKeyDown(event, node)
                            }
                            onClick={() => this.handleNodeClick(node)}
                            className={nodeButtonClasses}>
                            {showExpansionArrows && node.children ? (
                              <WuiIcon
                                className="wuiTreeView__expansionArrow"
                                size={display === 'compressed' ? 's' : 'm'}
                                type={
                                  this.isNodeOpen(node)
                                    ? 'arrowDown'
                                    : 'arrowRight'
                                }
                              />
                            ) : null}
                            {node.icon && !node.useEmptyIcon ? (
                              <span className="wuiTreeView__iconWrapper">
                                {this.isNodeOpen(node) && node.iconWhenExpanded
                                  ? node.iconWhenExpanded
                                  : node.icon}
                              </span>
                            ) : null}
                            {node.useEmptyIcon && !node.icon ? (
                              <span className="wuiTreeView__iconPlaceholder" />
                            ) : null}
                            <span className="wuiTreeView__nodeLabel">
                              {node.label}
                            </span>
                          </button>
                          <div
                            id={`wuiNestedTreeView-${this.state.treeID}`}
                            onKeyDown={(event: React.KeyboardEvent) =>
                              this.onChildrenKeydown(event, index)
                            }>
                            {node.children && this.isNodeOpen(node) ? (
                              <WuiTreeView
                                items={node.children}
                                display={display}
                                showExpansionArrows={showExpansionArrows}
                                expandByDefault={this.state.expandChildNodes}
                                {...label}
                              />
                            ) : null}
                          </div>
                        </li>
                      </React.Fragment>
                    );
                  }}
                </WuiI18n>
              );
            })}
          </ul>
        </WuiText>
      </WuiTreeViewContext.Provider>
    );
  }
}
