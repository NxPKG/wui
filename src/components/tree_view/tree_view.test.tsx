/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { WuiIcon } from '../icon';
import { WuiToken } from '../token';
import { render, shallow } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { WuiTreeView } from './tree_view';

// Mock the htmlIdGenerator to generate predictable ids for snapshot tests
jest.mock('../../services/accessibility/html_id_generator', () => ({
  htmlIdGenerator: () => () => 'htmlId',
}));

const items = [
  {
    label: 'Item One',
    id: 'item_one',
    icon: <WuiIcon type="folderClosed" />,
    iconWhenExpanded: <WuiIcon type="folderOpen" />,
    isExpanded: true,
    children: [
      {
        label: 'Item A',
        id: 'item_a',
        icon: <WuiIcon type="document" />,
      },
      {
        label: 'Item B',
        id: 'item_b',
        icon: <WuiIcon type="arrowRight" />,
        iconWhenExpanded: <WuiIcon type="arrowDown" />,
        children: [
          {
            label: 'A Cloud',
            id: 'item_cloud',
            icon: <WuiToken iconType="tokenConstant" />,
          },
          {
            label: "I'm a Bug",
            id: 'item_bug',
            icon: <WuiToken iconType="tokenEnum" />,
            className: 'classForBug',
          },
        ],
      },
      {
        label: 'Item C',
        id: 'item_c',
        icon: <WuiIcon type="arrowRight" />,
        iconWhenExpanded: <WuiIcon type="arrowDown" />,
        children: [
          {
            label: 'Another Cloud',
            id: 'item_cloud2',
            icon: <WuiToken iconType="tokenConstant" />,
          },
          {
            label: 'Another Bug',
            id: 'item_bug2',
            icon: <WuiToken iconType="tokenEnum" />,
          },
        ],
      },
    ],
  },
  {
    label: 'Item Two',
    id: 'item_two',
  },
];

describe('WuiTreeView', () => {
  test('is rendered', () => {
    const component = render(<WuiTreeView items={items} {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  test('length of open items', () => {
    const component = shallow<WuiTreeView>(
      <WuiTreeView items={items} {...requiredProps} />
    );
    const instance = component.instance();

    expect(component.state('openItems')).toHaveLength(1);

    instance.handleNodeClick(items[1]);
    expect(component.state('openItems')).toHaveLength(2);
  });

  test('activeItem changes', () => {
    const component = shallow<WuiTreeView>(
      <WuiTreeView items={items} {...requiredProps} />
    );
    const instance = component.instance();

    expect(component.state('activeItem')).toBe('');

    instance.handleNodeClick(items[1]);
    expect(component.state('activeItem')).toBe('item_two');
  });

  test('open node changes', () => {
    const component = shallow<WuiTreeView>(
      <WuiTreeView items={items} {...requiredProps} />
    );
    const instance = component.instance();

    expect(instance.isNodeOpen(items[1])).toBe(false);

    instance.handleNodeClick(items[1]);
    expect(instance.isNodeOpen(items[1])).toBe(true);

    expect(instance.isNodeOpen(items[0])).toBe(true);

    instance.handleNodeClick(items[0]);
    expect(instance.isNodeOpen(items[0])).toBe(false);
  });
});
