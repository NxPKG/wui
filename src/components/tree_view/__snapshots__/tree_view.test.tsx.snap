// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiTreeView is rendered 1`] = `
<div
  class="wuiText wuiText--medium wuiTreeView__wrapper"
>
  <p
    class="wuiScreenReaderOnly"
    id="htmlId--instruction"
  >
    You can quickly navigate this list using arrow keys.
  </p>
  <ul
    aria-describedby="htmlId--instruction"
    aria-label="aria-label"
    class="wuiTreeView testClass1 testClass2"
    data-test-subj="test subject string"
    id="htmlId"
  >
    <li
      class="wuiTreeView__node wuiTreeView__node--expanded"
    >
      <button
        aria-controls="wuiNestedTreeView-htmlId"
        aria-expanded="true"
        class="wuiTreeView__nodeInner"
        data-test-subj="wuiTreeViewButton-htmlId"
        id="htmlId--0--node"
      >
        <span
          class="wuiTreeView__iconWrapper"
        >
          <div
            data-wuiicon-type="folderOpen"
          />
        </span>
        <span
          class="wuiTreeView__nodeLabel"
        >
          Item One
        </span>
      </button>
      <div
        id="wuiNestedTreeView-htmlId"
      >
        <div
          class="wuiText wuiText--medium wuiTreeView__wrapper"
        >
          <ul
            aria-label="Item One child of aria-label"
            class="wuiTreeView"
            id="htmlId"
          >
            <li
              class="wuiTreeView__node"
            >
              <button
                aria-controls="wuiNestedTreeView-htmlId"
                aria-expanded="false"
                class="wuiTreeView__nodeInner"
                data-test-subj="wuiTreeViewButton-htmlId"
                id="htmlId--0--node"
              >
                <span
                  class="wuiTreeView__iconWrapper"
                >
                  <div
                    data-wuiicon-type="document"
                  />
                </span>
                <span
                  class="wuiTreeView__nodeLabel"
                >
                  Item A
                </span>
              </button>
              <div
                id="wuiNestedTreeView-htmlId"
              />
            </li>
            <li
              class="wuiTreeView__node"
            >
              <button
                aria-controls="wuiNestedTreeView-htmlId"
                aria-expanded="false"
                class="wuiTreeView__nodeInner"
                data-test-subj="wuiTreeViewButton-htmlId"
                id="htmlId--1--node"
              >
                <span
                  class="wuiTreeView__iconWrapper"
                >
                  <div
                    data-wuiicon-type="arrowRight"
                  />
                </span>
                <span
                  class="wuiTreeView__nodeLabel"
                >
                  Item B
                </span>
              </button>
              <div
                id="wuiNestedTreeView-htmlId"
              />
            </li>
            <li
              class="wuiTreeView__node"
            >
              <button
                aria-controls="wuiNestedTreeView-htmlId"
                aria-expanded="false"
                class="wuiTreeView__nodeInner"
                data-test-subj="wuiTreeViewButton-htmlId"
                id="htmlId--2--node"
              >
                <span
                  class="wuiTreeView__iconWrapper"
                >
                  <div
                    data-wuiicon-type="arrowRight"
                  />
                </span>
                <span
                  class="wuiTreeView__nodeLabel"
                >
                  Item C
                </span>
              </button>
              <div
                id="wuiNestedTreeView-htmlId"
              />
            </li>
          </ul>
        </div>
      </div>
    </li>
    <li
      class="wuiTreeView__node"
    >
      <button
        aria-controls="wuiNestedTreeView-htmlId"
        aria-expanded="false"
        class="wuiTreeView__nodeInner"
        data-test-subj="wuiTreeViewButton-htmlId"
        id="htmlId--1--node"
      >
        <span
          class="wuiTreeView__nodeLabel"
        >
          Item Two
        </span>
      </button>
      <div
        id="wuiNestedTreeView-htmlId"
      />
    </li>
  </ul>
</div>
`;
